ffi/src/clipboard/windows.rs

```diff
@@ -0,0 +1,123 @@
+pub mod ffi {
+
+    use crate::clipboard::ffi::CliprdrBackendFactory;
+    use crate::clipboard::message::ffi::ClipboardMessage;
+    use crate::error::ffi::IronRdpError;
+
+    use super::WinCliprdrInner;
+
+    #[diplomat::opaque]
+    pub struct WinCliprdr(WinCliprdrInner);
+
+    impl WinCliprdr {
+        pub fn new(hwnd: isize) -> Result<Box<WinCliprdr>, Box<IronRdpError>> {
+            WinCliprdrInner::new(hwnd).map(WinCliprdr).map(Box::new)
+        }
+
+        pub fn next_clipboard_message(&self) -> Result<Option<Box<ClipboardMessage>>, Box<IronRdpError>> {
+            Ok(self.0.next_clipboard_message()?.map(ClipboardMessage).map(Box::new))
+        }
+
+        pub fn next_clipboard_message_blocking(&self) -> Result<Box<ClipboardMessage>, Box<IronRdpError>> {
+            self.0
+                .next_clipboard_message_blocking()
+                .map(ClipboardMessage)
+                .map(Box::new)
+        }
+
+        pub fn backend_factory(&self) -> Result<Box<CliprdrBackendFactory>, Box<IronRdpError>> {
+            self.0.backend_factory().map(Box::new)
+        }
+    }
+}
+
+#[cfg(not(windows))]
+pub struct WinCliprdrInner;
+
+#[cfg(not(windows))]
+impl WinCliprdrInner {
+    fn new(_hwnd: isize) -> Result<WinCliprdrInner, Box<IronRdpError>> {
+        Err(WrongOSError::expected_platform("windows")
+            .with_custom_message("WinCliprdr only support windows")
+            .into())
+    }
+
+    fn next_clipboard_message(&self) -> Result<Option<ironrdp::cliprdr::backend::ClipboardMessage>, Box<IronRdpError>> {
+        Err(WrongOSError::expected_platform("windows")
+            .with_custom_message("WinCliprdr only support windows")
+            .into())
+    }
+
+    fn backend_factory(&self) -> Result<CliprdrBackendFactory, Box<IronRdpError>> {
+        Err(WrongOSError::expected_platform("windows")
+            .with_custom_message("WinCliprdr only support windows")
+            .into())
+    }
+
+    fn next_clipboard_message_blocking(
+        &self,
+    ) -> Result<ironrdp::cliprdr::backend::ClipboardMessage, Box<IronRdpError>> {
+        Err(WrongOSError::expected_platform("windows")
+            .with_custom_message("WinCliprdr only support windows")
+            .into())
+    }
+}
+
+#[cfg(windows)]
+pub struct WinCliprdrInner {
+    pub clipboard: ironrdp_cliprdr_native::WinClipboard,
+    pub receiver: std::sync::mpsc::Receiver<ironrdp::cliprdr::backend::ClipboardMessage>,
+}
+
+#[cfg(windows)]
+impl WinCliprdrInner {
+    fn new(hwnd: isize) -> Result<WinCliprdrInner, Box<IronRdpError>> {
+        use windows::Win32::Foundation::HWND;
+
+        let (sender, receiver) = std::sync::mpsc::channel();
+
+        let proxy = crate::clipboard::FfiClipbarodMessageProxy { sender };
+
+        // SAFETY: `hwnd` must be a valid window handle
+        let clipboard = unsafe { ironrdp_cliprdr_native::WinClipboard::new(HWND(hwnd), proxy) }?;
+
+        Ok(WinCliprdrInner { clipboard, receiver })
+    }
```

Comment:
issue (non-blocking): The `SAFETY:` comment above the unsafe usage should explain why the preconditions are upheld. This comment is explaining what is the precondition, but there is no guarantee it is upheld as the value of `hwnd` is arbitrary. In this case, the function should be made unsafe and the comment be part of the safety section documenting the function. If Diplomat does not allow the function to be marked unsafe, let’s at least keep the comment at the right place.

Reply:
Added comments, I get errors from compiler with unsafe function in FFI block.